File: src/lib/api/wordlist.ts
/**
 * Wordlist API functions
 */

import { api } from '../api';

// Types matching backend schemas
export interface WordlistEntry {
    id: number;
    word: string;
    definition: string;
    part_of_speech: string;
    difficulty?: number;
    added_at: string;
    last_reviewed?: string;
    review_count: number;
    mastery_level: number;
    notes?: string;
    full_definition?: Record<string, unknown>;
}

export interface WordlistResponse {
    total: number;
    words: WordlistEntry[];
}

export interface WordlistStats {
    total_words: number;
    words_mastered: number;
    words_learning: number;
    words_new: number;
    average_mastery: number;
}

export interface AddWordRequest {
    word: string;
    notes?: string;
}

export interface UpdateWordRequest {
    notes?: string;
    mastery_level?: number;
}

export interface RandomWordsRequest {
    count?: number;
    topic?: string;
    min_mastery?: number;
    max_mastery?: number;
}

export interface RandomWord {
    word: string;
    definition: string;
    part_of_speech: string;
    mastery_level: number;
}

/**
 * Get user's complete word list
 */
export async function getUserWordlist(): Promise<WordlistResponse> {
    return api.get<WordlistResponse>('/api/wordlist');
}

/**
 * Add a word to user's word list
 */
export async function addToWordlist(word: string, notes?: string): Promise<WordlistEntry> {
    return api.post<WordlistEntry>('/api/wordlist', { word, notes });
}

/**
 * Remove a word from user's word list
 */
export async function removeFromWordlist(word: string): Promise<void> {
    await api.delete(`/api/wordlist/${encodeURIComponent(word)}`);
}

/**
 * Update a word list entry
 */
export async function updateWordlistEntry(
    word: string,
    data: UpdateWordRequest
): Promise<WordlistEntry> {
    return api.patch<WordlistEntry>(`/api/wordlist/${encodeURIComponent(word)}`, data);
}

/**
 * Get wordlist statistics
 */
export async function getWordlistStats(): Promise<WordlistStats> {
    return api.get<WordlistStats>('/api/wordlist/stats');
}

/**
 * Get random words from wordlist for practice
 */
export async function getRandomWords(
    options?: RandomWordsRequest
): Promise<{ words: RandomWord[] }> {
    return api.post<{ words: RandomWord[] }>('/api/wordlist/random', options || {});
}

/**
 * Check if a word is in user's wordlist
 */
export async function isWordInList(wordlist: WordlistEntry[], word: string): boolean {
    return wordlist.some(entry => entry.word.toLowerCase() === word.toLowerCase());
}

export default {
    getUserWordlist,
    addToWordlist,
    removeFromWordlist,
    updateWordlistEntry,
    getWordlistStats,
    getRandomWords,
    isWordInList,
};

File: backend/app/schemas/wordlist.py
"""Pydantic schemas for wordlist API."""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class WordlistEntryBase(BaseModel):
    """Base schema for wordlist entry."""

    notes: Optional[str] = Field(None, description="User's personal notes for this word")


class WordlistAddRequest(BaseModel):
    """Request to add a word to user's wordlist."""

    word: str = Field(..., min_length=1, max_length=100, description="Word to add")
    notes: Optional[str] = Field(None, description="Optional notes")


class WordlistUpdateRequest(BaseModel):
    """Request to update a wordlist entry."""

    notes: Optional[str] = Field(None, description="Updated notes")
    mastery_level: Optional[int] = Field(None, ge=0, le=100, description="Updated mastery level")


class WordlistEntryResponse(BaseModel):
    """Response for a single wordlist entry."""

    id: int
    word: str
    definition: str
    part_of_speech: str
    difficulty: Optional[int] = None
    added_at: datetime
    last_reviewed: Optional[datetime] = None
    review_count: int
    mastery_level: int
    notes: Optional[str] = None

    class Config:
        from_attributes = True


class WordlistResponse(BaseModel):
    """Response for user's wordlist."""

    total: int = Field(..., description="Total number of words in list")
    words: list[WordlistEntryResponse] = Field(..., description="List of words")


class WordlistStatsResponse(BaseModel):
    """Response for wordlist statistics."""

    total_words: int
    words_mastered: int  # mastery_level >= 80
    words_learning: int  # mastery_level 20-79
    words_new: int  # mastery_level < 20
    average_mastery: float


class RandomWordsRequest(BaseModel):
    """Request for random words from wordlist."""

    count: int = Field(5, ge=1, le=10, description="Number of words to return")
    topic: Optional[str] = Field(None, description="Optional topic for relevance filtering")
    min_mastery: Optional[int] = Field(None, ge=0, le=100, description="Minimum mastery level")
    max_mastery: Optional[int] = Field(None, ge=0, le=100, description="Maximum mastery level")
